### PR TITLE
How to use PSRAM with ESP32 on Zephyr

### PR DESCRIPTION
**Pull request by wmrsouza**
This article aims to show how to enable, configure and use external PSRAM memory on Zephyr applications running on ESP32 SoCs

### CODE REVIEW

File: content/blog/zephyr-how-to-use-psram/index.md [hunk 1]
```diff
@@ -0,0 +1,605 @@
+---
+title: "Enabling External PSRAM for Zephyr Applications on ESP32 SoCs"
+date: 2024-12-04T00:00:00-00:00
+showAuthor: false
+featureAsset: "./psram_cover.webp"
```
Comment (pedrominatel) [inline content/blog/zephyr-how-to-use-psram/index.md:5]: You can remove this line if you rename the `psram_cover.webp` file to `featured.webp`.

File: content/blog/zephyr-how-to-use-psram/index.md [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+
+```
+
+Here, we can confirm that instructions and read-only data were copied and mapped to PSRAM, which can optimize RAM usage when there is code that needs to execute during an SPI1 Flash operation.
+
+## 9. Conclusion
+
+Throughout this article, we explored three different strategies for utilizing PSRAM: using it as a heap for dynamic memory allocation, integrating it as a section in the memory map, and moving code from FLASH to PSRAM. In addition to these three strategies, it is also possible to leverage PSRAM to execute external code compiled as Position Independent Code (PIC), which can be placed in PSRAM by a loader for execution. Another potential use of PSRAM to reduce SRAM usage is by enabling **CONFIG_ESP32_WIFI_NET_ALLOC_SPIRAM**, allowing the ESP32 Wi-Fi stack to dynamically allocate memory from PSRAM. However, as of the time of writing, there is an unfixed bug preventing applications built with **CONFIG_ESP32_WIFI_NET_ALLOC_SPIRAM** from functioning properly.
+
+## 10. References
```
Comment (pedrominatel) [inline content/blog/zephyr-how-to-use-psram/index.md:581]: For the references, please add the links using the Markdown format:

`[Espressif Official Site](https://www.espressif.com/en)`

File: content/blog/zephyr-how-to-use-psram/index.md [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+To know which ESP32 modules contain PSRAM we have to undestand how modules part nummber are formed. Lets use as an example the moduel whose part number is **ESP32-S3N8R8**
+
+| Field           | Contents | Meaning             |
+|:---------------:|:--------:|:-------------------:|
+| SoC series      | ESP32-S3 | ESP32-S3 SoC series |
+| FLASH size      | N8       | 8 MBytes FLASH      |
+| PSRAM size [^1] | R8       | 8 MBytes PSRAM      |
+
+ [^1]: If the module does not contain PSRAM, this field is absent
+
```
Comment (pedrominatel) [inline content/blog/zephyr-how-to-use-psram/index.md:38]: Remove the extra line.

File: content/blog/zephyr-how-to-use-psram/index.md [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+Before we start, let’s first understand exactly what PSRAM stands for and what it really is. The acronym PSRAM stands for pseudo-static random-access memory. It is a type of dynamic random-access memory (DRAM) that internally manages memory cell refresh, allowing it to function similarly to static random-access memory (SRAM) while retaining the cost and characteristics of DRAM. PSRAM is typically accessible via high-speed serial buses, and some ESP32 series SoCs can communicate with PSRAM and map its contents into the CPU's virtual memory space through a memory management unit (MMU).
+
+In a simplified way, we can describe the relationship among PSRAM, the MMU, and the CPU as follows:
+
+- Data is automatically copied from PSRAM to the data cache when the CPU attempts to access a memory position not currently in the cache.
+
+- Conversely, when CPU modifies data in the cache, the updated contents are copied back to PSRAM.
+
+As of the time of writing, although ESP32, ESP32-S2, ESP32-S3, ESP32-C61, and ESP32-P4 support external PSRAM, only the former three are currently supported on Zephyr.
```
Comment (pedrominatel) [inline content/blog/zephyr-how-to-use-psram/index.md:25]: Maybe here you can be more explicit about the SoCs that are supported on Zephyr.

File: content/blog/zephyr-how-to-use-psram/index.md [hunk 5]
```diff
@@ -0,0 +1,605 @@
...
+The following SPI modes are supported to access external PSRAM memories:
+
+ - Single SPI
+ - Dual SPI
+ - Quad SPI
+ - QPI
+ - Octo SPI
+ - OPI
+
+ESP32 can manage PSRAM of capacities up 32Mbytes depending on  SoCs series.
```
Comment (pedrominatel) [inline content/blog/zephyr-how-to-use-psram/index.md:55]: 32 Mbits (ESP32 max support) and 32 MBytes (ESP32-S3).
This could be confusing if you don't specify the ESP32 you are talking about or if you are mixing bits with Bytes. I recommend you removing and just keep it as:

`Espressif currently some PSRAM models:`

File: content/blog/zephyr-how-to-use-psram/index.md [hunk 6]
```diff
@@ -0,0 +1,605 @@
...
+| ESP-PSRAM64  | 64Mbit          | 1.8V               |
+| ESP-PSRAM64H | 64Mbit          | 3.3V               |
+
+## 4. Configuring PSRAM for use in Zephyr
+
+The Kconfig parameters related to PSRAM usage in Zephyr can be found in [zephyr/soc/espressif/commom/Kconfig.spiram](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/espressif/common/Kconfig.spiram)
+
+Below are the main parameters related to PSRAM, along with a brief description of each.
+
+### 4.1. ESP_SPIRAM
```
Comment (pedrominatel) [inline content/blog/zephyr-how-to-use-psram/index.md:72]: To reduce the ToC, please use:

`- ESP_SPIRAM` instead of `### 4.1. ESP_SPIRAM`

File: content/blog/zephyr-how-to-use-psram/index.md [hunk 7]
```diff
@@ -0,0 +1,605 @@
...
+
+This configuration parameter enables support for an external SPI RAM chip, connected in parallel with the main SPI flash chip. If enabled it automatically enables CONFIG_SHARED_MULTI_HEAP.
+
+### 4.2. ESP_HEAP_SEARCH_ALL_REGIONS
```
Comment (pedrominatel) [inline content/blog/zephyr-how-to-use-psram/index.md:76]: Same for all items from 4.x.

File: content/blog/zephyr-how-to-use-psram/index.md [hunk 8]
```diff
@@ -0,0 +1,605 @@
...
+ - Quad SPI
+ - QPI
+ - Octo SPI
+ - OPI
+
+ESP32 can manage PSRAM of capacities up 32Mbytes depending on  SoCs series.
+
+Espressif currently some PSRAM models:
+
+| Part number  | Memory Capacity | Operationg Voltage |
```
Comment (pedrominatel) [inline content/blog/zephyr-how-to-use-psram/index.md:59]: Sometimes you're using Mb and MB. I recommend using the same units to avoid confusion.

File: content/blog/zephyr-how-to-use-psram/index.md [hunk 9]
```diff
@@ -0,0 +1,605 @@
+---
```
Comment (pedrominatel) [inline content/blog/zephyr-how-to-use-psram/index.md:1]: Move the article folder to `/content/blog/2024/12/`.

File: content/blog/zephyr-how-to-use-psram/index.md [hunk 10]
```diff
@@ -0,0 +1,605 @@
...
+
+	return 0;
+}
+```
+
+To build a project containing these two files targeting a **ESP32S3_DevKitC** board:
+
+```sh
+$ west build -b esp32s3_devkitc/esp32s3/procpu <project folder path> --pristine
+```
```
Comment (pedrominatel) [inline content/blog/zephyr-how-to-use-psram/index.md:337]: Add an extra line here.

File: content/blog/zephyr-how-to-use-psram/index.md [hunk 11]
```diff
@@ -0,0 +1,605 @@
...
+10094  *fill*	      0x000000003c0a0000    0x80000
+10095   	      0x000000003c120000		. = ALIGN (0x4)
+10096   	      0x000000003c120000		_ext_ram_bss_end = ABSOLUTE (.)
+
+...
+```
+
+Here we can ovserve that first position of **psram_vector** is at addres **0x3c020000** which is inside the region mapping external PSRAM on ESP32-C3. Beyond that we can observe that although **CONFIG_SHARED_MULTI_HEAP** was not explicitaly enable, it was reserved area for spiram_head because **CONFIG_SHARED_MULTI_HEAP** is enabled by defaul once **CONFIG_ESP_SPIRAM** is enabled.
+
+Now let's flash the binary onto the **ESP32S3_DevKitC** board and observe the messages from the sample:
```
Comment (pedrominatel) [inline content/blog/zephyr-how-to-use-psram/index.md:362]: ESP32-S3-DevKitC-1

File: content/blog/zephyr-how-to-use-psram/index.md [hunk 12]
```diff
@@ -0,0 +1,605 @@
...
+[00:00:00.373,000] <inf> PSRAM_SAMPLE: Sample finished successfully!
+```
+
+Here, we can observe that the application now starts earlier compared to the previous sample. This improvement results from disabling the **CONFIG_ESP_SPIRAM_MEMTEST** parameter, which bypasses the PSRAM memory test that previously took a few hundred milliseconds.
+
+## 8. Moving code from FLASH to PSRAM
+
+Enabling **CONFIG_SPIRAM_FETCH_INSTRUCTIONS** will move instructions from FLASH to PSRAM during startup, and if **CONFIG_SPIRAM_RODATA** is also enabled, code that requires execution during an SPI1 Flash operation can forgo being placed in IRAM, thus optimizing RAM usage.
+
+To check the effects of **CONFIG_SPIRAM_FETCH_INSTRUCTIONS** and **CONFIG_SPIRAM_RODATA** let's build **hello_world** firts withou enabling such parameters and then enable them.
```
Comment (pedrominatel) [inline content/blog/zephyr-how-to-use-psram/index.md:432]: "firts withou" -> "first without"

**Review by pedrominatel — COMMENTED** [review]
Thank you so much, @wmrsouza for your contribution!
PTAL at my comments.

**Comment by f-hollow** [comment]
@wmrsouza Please start by reviewing my PR to your branch `zephyr_how_to_use_psram` with some spelling and grammar fixes. It should save your time implementing my review comments.

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 13]
```diff
@@ -0,0 +1,616 @@
...
+
+Before we start, let’s first understand exactly what PSRAM stands for and what it really is. The acronym PSRAM stands for pseudo-static random-access memory. It is a type of dynamic random-access memory (DRAM) that internally manages memory cell refresh, allowing it to function similarly to static random-access memory (SRAM) while retaining the cost and characteristics of DRAM. PSRAM is typically accessible via high-speed serial buses, and some ESP32 series SoCs can communicate with PSRAM and map its contents into the CPU's virtual memory space through a memory management unit (MMU).
+
+In a simplified way, we can describe the relationship among PSRAM, the MMU, and the CPU as follows:
+
+- Data is automatically copied from PSRAM to the data cache when the CPU attempts to access a memory position not currently in the cache.
+
+- Conversely, when CPU modifies data in the cache, the updated contents are copied back to PSRAM.
+
+As of the time of writing, although ESP32, ESP32-S2, ESP32-S3, ESP32-C61, and ESP32-P4 support external PSRAM, ESP32-C6 and ESP32-P4 are not currently supported on Zephyr yet.
```
[the remark this answers is not in the document]
Comment (wmrsouza) [reply content/blog/2024/12/zephyr-how-to-use-psram/index.md:24]: ESP32-C61

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 14]
```diff
@@ -0,0 +1,616 @@
...
+*** Booting Zephyr OS build v4.0.0-1471-ge60f04096cd8 ***
+[00:00:00.301,000] <inf> PSRAM_SAMPLE: Sample started
+[00:00:00.373,000] <inf> PSRAM_SAMPLE: Sample finished successfully!
+```
+
+Here, we can observe that the application now starts earlier compared to the previous sample. This improvement results from disabling the **CONFIG_ESP_SPIRAM_MEMTEST** parameter, which bypasses the PSRAM memory test that previously took a few hundred milliseconds.
+
+## 8. Moving code from FLASH to PSRAM
+
+Enabling **CONFIG_SPIRAM_FETCH_INSTRUCTIONS** will move instructions from FLASH to PSRAM during startup, and if **CONFIG_SPIRAM_RODATA** is also enabled, code that requires execution during an SPI1 Flash operation can forgo being placed in IRAM, thus optimizing RAM usage.
```
[the remark this answers is not in the document]
Comment (wmrsouza) [reply content/blog/2024/12/zephyr-how-to-use-psram/index.md:432]: I'll try to address this matter keeping the "parameter name" along the explanations and use `CONFIG_` in sample code as user will have to do.

**Comment by marekmatej** [comment]
@wmrsouza please rebase to enjoy the article preview link being created by the bot.

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 15]
```diff
@@ -0,0 +1,587 @@
...
+date: 2024-12-04T00:00:00-00:00
+showAuthor: false
+authors:
+  - "marcio-ribeiro"
+tags: ["ESP32", "ESP32-S2", "ESP32-S3", "Zephyr", "PSRAM", "SPIRAM"]
+---
+
+Although ESP32 SoCs typically feature a few hundred kilobytes of internal SRAM, some embedded IoT applications—such as those featuring sophisticated graphical user interfaces, handling large data volumes, or performing complex data processing—may require much more RAM. To meet these demands, Espressif provides modules that feature external PSRAM.
+
+In this article, you will learn how to configure Zephyr to enable your application take full advantage of the PSRAM available on ESP32 modules. We will explore three different strategies for utilizing this PSRAM: using it as a heap for dynamic memory allocation, integrating it as a section into the memory map, and moving code from flash to PSRAM.
```
Comment (raffarost) [inline content/blog/2024/12/zephyr-how-to-use-psram/index.md:12]: to enable your application `to` take...

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 16]
```diff
@@ -0,0 +1,587 @@
...
+
+Before we start, let’s first understand exactly what PSRAM stands for and what it really is. The acronym PSRAM stands for pseudo-static random-access memory. It is a type of dynamic random-access memory (DRAM) that internally manages memory cell refresh, allowing it to function similarly to static random-access memory (SRAM) while retaining the cost and characteristics of DRAM. PSRAM is typically accessible via high-speed serial buses, and some ESP32 series SoCs can communicate with PSRAM and map its contents into the CPU's virtual memory space through a memory management unit (MMU).
+
+In a simplified way, we can describe the relationship among PSRAM, the MMU, and the CPU as follows:
+
+- Data is automatically copied from PSRAM to the data cache when the CPU attempts to access a memory position not currently in the cache.
+
+- Conversely, when CPU modifies data in the cache, the updated contents are copied back to PSRAM.
+
+As of the time of writing, although ESP32, ESP32-S2, ESP32-S3, ESP32-C61, and ESP32-P4 support external PSRAM, but ESP32-C61 and ESP32-P4 are not currently supported on Zephyr yet.
```
Comment (raffarost) [inline content/blog/2024/12/zephyr-how-to-use-psram/index.md:24]: I found the phrase a bit confusing, specially the word `although`
Maybe rephrase to become clear which HW supports PSRAM and which devices are supported in Zephyr?

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 17]
```diff
@@ -0,0 +1,587 @@
...
+The following SPI modes are supported to access external PSRAM memories:
+
+ - Single SPI
+ - Dual SPI
+ - Quad SPI
+ - QPI
+ - Octo SPI
+ - OPI
+
+The are limits regarding the maximum PSRAM capacities that each ESP32 SoC series can manage. As an example, ESP32-S3 can manage PSRAMs up to 32 MB.
```
Comment (raffarost) [inline content/blog/2024/12/zephyr-how-to-use-psram/index.md:55]: The ->There
capacities -> capacity

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 18]
```diff
@@ -0,0 +1,587 @@
...
+***SPIRAM_SPEED:*** This configuration parameter sets the SPIRAM clock speed in MHz:
+ | Value | Clock speed |
+ | --- | --- |
+ | SPIRAM_SPEED_20M  |  20 MHz  |
+ | SPIRAM_SPEED_26M  |  26 MHz  |
+ | SPIRAM_SPEED_40M  |  40 MHz  |
+ | SPIRAM_SPEED_80M  |  80 MHz  |
+ | SPIRAM_SPEED_120M | 120 MHz  |
+
+***SPIRAM_FETCH_INSTRUCTIONS:*** This configuration parameter allows moving Instructions from flash to PSRAM. If enabled, instructions in flash will be moved into PSRAM on startup. If ```SPIRAM_RODATA``` parameter is also enabled, the code that normally requires execution during the SPI1 flash operation does not need to be placed in IRAM, thus optimizing RAM usage. By default, this parameter is disabled.
```
Comment (raffarost) [inline content/blog/2024/12/zephyr-how-to-use-psram/index.md:99]: Instructions -> instructions (maybe no cap)

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 19]
```diff
@@ -0,0 +1,587 @@
...
+
+***SPIRAM_RODATA:*** This configuration parameter allows moving read-only data from flash to PSRAM. If ```SPIRAM_FETCH_INSTRUCTIONS``` parameter is also enabled, the code that normally requires execution during the SPI1 flash operation  does not need to be placed in IRAM, thus optimizing RAM usage.
```
Comment (raffarost) [inline content/blog/2024/12/zephyr-how-to-use-psram/index.md:101]: operation  does -> operation does (double spacing)

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 20]
```diff
@@ -0,0 +1,587 @@
...
+
+## Installing Zephyr and its dependencies
+
+To install Zephyr RTOS and the necessary tools, follow the instructions in the Zephyr [Getting Started Guide](https://docs.zephyrproject.org/latest/develop/getting_started/index.html). By the end of the process, you will have a command-line Zephyr development environment set up and ready to build your application with `west` — the meta-tool responsible for building your application and flashing the generated binary, and other tasks.
```
Comment (raffarost) [inline content/blog/2024/12/zephyr-how-to-use-psram/index.md:105]: instructions in the Zephyr -> in Zephyr's (suggestion only)

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 21]
```diff
@@ -0,0 +1,587 @@
...
+
+Additionally, you need to execute the following command to prepare your environment for building applications for Espressif SoCs:
+
+```sh
+west blobs fetch hal_espressif
+```
+
+## Dynamically Allocating PSRAM Memory
+
+PSRAM memory blocks can be made available to applications through Zephyr's shared multi-heap library.  The shared multi-heap memory pool manager uses the multi-heap allocator to manage a set of reserved memory regions with varying capabilities and attributes. For PSRAM, enabling the ```ESP_SPIRAM``` and ```SHARED_MULTI_HEAP``` parameters causes the external PSRAM to be mapped into the data virtual memory space during Zephyr's early initialization stage. The shared multi-heap framework is initialized, and the PSRAM memory region is added to the pool.
```
Comment (raffarost) [inline content/blog/2024/12/zephyr-how-to-use-psram/index.md:115]: library.  The (double spacing)
maybe use the editor tool for a search, review and replace is an easy fix
there are more cases below

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 22]
```diff
@@ -0,0 +1,587 @@
...
+	return 0;
+}
+```
+
+To build a project containing these two files targeting a `ESP32-S3-DevKitC-1` board:
+
+```sh
+west build -b esp32s3_devkitc/esp32s3/procpu <project folder path> --pristine
+```
+This command will create a directory called `build`, which will contain the binary file of our sample along with other intermediate files produced during the building process
```
Comment (raffarost) [inline content/blog/2024/12/zephyr-how-to-use-psram/index.md:187]: add period `.` at the end of phrase

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 23]
```diff
@@ -0,0 +1,587 @@
...
+*** Booting Zephyr OS build v4.0.0-1471-ge60f04096cd8 ***
+[00:00:00.708,000] <inf> PSRAM_SAMPLE: Sample started
+[00:00:00.708,000] <inf> PSRAM_SAMPLE: Sample finished successfully!
+```
+
+Once the sample finishes executing successfully, we can conclude that the memory allocated from PSRAM was read and written to without any issues.
+
+## Integrating PSRAM memory as a section into the memory map
+
+Once ```ESP_SPIRAM``` is enabled, a section called `.ext_ram.bss` will be created. This section will hold non-initialized global variables that will later be placed in PSRAM. These gloabal variables must be declared with `__attribute__ ((section (".ext_ram.bss"))`.
```
Comment (raffarost) [inline content/blog/2024/12/zephyr-how-to-use-psram/index.md:264]: gloabal -> global

**Review by raffarost — APPROVED** [review]
nice and informative, LGTM

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 24]
```diff
@@ -0,0 +1,741 @@
...
+
+Here, we can observe that the application now starts earlier compared to the previous sample. This improvement results from disabling the ```ESP_SPIRAM_MEMTEST``` parameter, which bypasses the PSRAM memory test that previously took a few hundred milliseconds.
+
+## Running Code from PSRAM Instead of FLASH
+
+Enabling the ```SPIRAM_FETCH_INSTRUCTIONS``` parameter will move instructions from FLASH to PSRAM during startup, and if the ```ESP_SPIRAM_MEMTEST``` parameter is also enabled, code that requires execution during an SPI1 Flash operation can forgo being placed in IRAM, thus optimizing RAM usage.
+
+To check the effects of ```SPIRAM_FETCH_INSTRUCTIONS``` and ```SPIRAM_RODATA``` parameters, let's build `hello_world` first without enabling them and then enable them.
+
+***Building `hello_world` with `SPIRAM_FETCH_INSTRUCTIONS` and `SPIRAM_RODATA` disbled***
```
Comment (f-hollow) [inline content/blog/2024/12/zephyr-how-to-use-psram/index.md:419]:
```suggestion
***Building `hello_world` with `SPIRAM_FETCH_INSTRUCTIONS` and `SPIRAM_RODATA` disabled***
```

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 25]
```diff
@@ -0,0 +1,741 @@
...
+
+## Letting ESP32 Wi-Fi Use PSRAM
+
+Another potential use of PSRAM to reduce SRAM usage is by enabling `ESP32_WIFI_NET_ALLOC_SPIRAM` parameter, allowing the ESP32 Wi-Fi stack to dynamically allocate memory from PSRAM.
+
+To check the results, let's build the `zephyr/samples/net/wifi/shell`, first with `ESP32_WIFI_NET_ALLOC_SPIRAM` disabled and then with this parameter enabled. In both cases let's flash the binary into the board and launch a console to interact with the shell. After connect to an access point we will execute `net allocs` to get information from where memory is being allocated in both cases.
+
+***Building `zephyr/samples/net/wifi/shell` with `ESP32_WIFI_NET_ALLOC_SPIRAM` disabled***
+
+Building, flashing and monitoring:
```
Comment (f-hollow) [inline content/blog/2024/12/zephyr-how-to-use-psram/index.md:561]:
```suggestion
Building, flashing, and monitoring:
```

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 26]
```diff
@@ -0,0 +1,741 @@
...
+28 bytes from 192.168.15.17 to 192.168.15.2: icmp_seq=3 ttl=64 time=338 ms
+uart:~$ net allocs
+Network memory allocations
+memory		Status	Pool	Function alloc -> freed
+0x3c083920	 free	   RX	eth_esp32_rx():135 -> processing_data():178
+0x3c0836a0	 free	   TX	dhcpv4_create_message():287 -> ethernet_send():762
+0x3c082894	 free	TDATA	ethernet_fill_header():532 -> ethernet_remove_l2_header():61
+```
+
+Examining the output from both sessions, we can confirm that in the first case, network stack allocations were made using memory from internal RAM -- addresses 0x3fca4af0, 0x3fca4870, and 0x3fca3a2c. We can also confirm that in the second case, network stack allocations were taking memory from PSRAM -- addresses 0x3c083920, 0x3c0836a0, and 0x3c082894 -- thus proving that enabling `ESP32_WIFI_NET_ALLOC_SPIRAM` avoids allocating memory from the precious internal SRAM and takes advantage of PSRAM.
```
Comment (f-hollow) [inline content/blog/2024/12/zephyr-how-to-use-psram/index.md:714]:
```suggestion
Examining the output from both sessions, we can confirm that in the first case, network stack allocations were made using memory from internal RAM -- addresses 0x3fca4af0, 0x3fca4870, and 0x3fca3a2c. We can also confirm that in the second case, network stack allocations were taking memory from PSRAM -- addresses `0x3c083920`, `0x3c0836a0`, and `0x3c082894` -- thus proving that enabling `ESP32_WIFI_NET_ALLOC_SPIRAM` avoids allocating memory from the precious internal SRAM and takes advantage of PSRAM.
```

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 27]
```diff
@@ -0,0 +1,741 @@
...
+
+## Final Thoughts
+
+Throughout this article, we explored three different strategies for utilizing PSRAM: using it as a heap for dynamic memory allocation, integrating it as a section in the memory map, and moving code from flash to PSRAM.
```
Comment (f-hollow) [inline content/blog/2024/12/zephyr-how-to-use-psram/index.md:718]: In the intro and here, you talk about the three strategies as this:

- using it as a heap for dynamic memory allocation
- integrating it as a section in the memory map
- moving code from flash to PSRAM

However, the actual article sections are called:

- Using PSRAM for Dynamic Memory Allocation
- Adding PSRAM to Your ESP32 Memory Map
- Running Code from PSRAM Instead of Flash

Suggest that you use the same names for these strategies both in text and section titles.

File: content/blog/2024/12/zephyr-how-to-use-psram/index.md [hunk 28]
```diff
@@ -0,0 +1,741 @@
...
+
+In addition to these three strategies, it is also possible to leverage PSRAM to execute external code compiled as Position Independent Code (PIC), which can be placed in PSRAM by a loader for execution. You can find more information regarding to this way of taking avantage of PSRAM consulting Zephyr's documentation about [LLEXT -- Linkable Loadable Extensions](https://docs.zephyrproject.org/latest/services/llext/index.html)
```
Comment (f-hollow) [inline content/blog/2024/12/zephyr-how-to-use-psram/index.md:720]:
```suggestion
In addition to these three strategies, it is also possible to leverage PSRAM to execute external code compiled as Position Independent Code (PIC), which can be placed in PSRAM by a loader for execution. You can find more information about this way of using PSRAM by consulting Zephyr's document [LLEXT -- Linkable Loadable Extensions](https://docs.zephyrproject.org/latest/services/llext/index.html).
```

**Review by f-hollow — APPROVED** [review]
Some final nitpicks and suggestions. Otherwise, LGTM!

**Review by pedrominatel — APPROVED** [review]
